### PR TITLE
Merge staged BoringSSL commits in `current-upstream-merge-20211128`

### PR DESCRIPTION
**Pull request by nebeid**
### Description of changes: 
Merge BorinSSL commits between Nov 15, starting at 27a67557405dcf18567f4528aa537da25f0a34e2  (https://github.com/google/boringssl/commit/27a67557405dcf18567f4528aa537da25f0a34e2) until Jan 10, ending at 0354b79d70ba9c8fdfd59a7a9cda1f02758688a5 (https://github.com/google/boringssl/commit/0354b79d70ba9c8fdfd59a7a9cda1f02758688a5).

See PR #366 "Boringssl merge 0354b79" for details on the changes, call-outs and benchmarking. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -270,6 +270,10 @@ if(GCC OR CLANG)
     set(C_CXX_FLAGS "${C_CXX_FLAGS} -Wimplicit-fallthrough")
   endif()
 
+  if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "5")
```
Comment (samuel40791765) [inline CMakeLists.txt:273]: Not sure if `AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "5"`should be added? I don't see it in the Boringssl `CMakeLists.txt` and it didn't show up in my merge attempts.

**Review by samuel40791765 — COMMENTED** [review]
One small comment found from doing the runbook, other stuff looks all good to me

File: CMakeLists.txt (hunk 1 above)
Comment (nebeid) [reply CMakeLists.txt:273]: Blame-ing on this file shows it comes from this commit https://github.com/awslabs/aws-lc/commit/f22f88ecb1e2afb7a0efaa0da23bbaecdd3f850f by @bryce-shang

Comment (bryce-shang) [reply CMakeLists.txt:273]: The flag does not exist before g++-5.

File: crypto/fipsmodule/cpucap/cpu_ppc64le.c [hunk 2]
```diff
@@ -12,7 +12,7 @@
  * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
  * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
 
-#include <openssl/cpu.h>
+#include <openssl/base.h>
 
 #if defined(OPENSSL_PPC64LE)
```
Comment (dkostic) [inline crypto/fipsmodule/cpucap/cpu_ppc64le.c:18]: do we need to include `../../internal.h` here?

File: crypto/internal.h [hunk 3]
```diff
@@ -952,6 +952,138 @@ void boringssl_fips_inc_counter(enum fips_counter_t counter);
...
+// bits in XCR0, so it is not necessary to check those.
+extern uint32_t OPENSSL_ia32cap_P[4];
+
+#if defined(BORINGSSL_FIPS) && !defined(BORINGSSL_SHARED_LIBRARY)
+const uint32_t *OPENSSL_ia32cap_get(void);
+#else
+OPENSSL_INLINE const uint32_t *OPENSSL_ia32cap_get(void) {
+  return OPENSSL_ia32cap_P;
+}
+#endif
```
Comment (dkostic) [inline crypto/internal.h:986]: This should be removed. We had to remove the `get` function for FIPS (Atsec insisted), it was done in #330

Comment (dkostic) [reply crypto/internal.h:986]: Lines 980-986, to be more precise

File: include/openssl/arm_arch.h [hunk 4]
```diff
@@ -117,8 +75,28 @@
 // ARMV8_PMULL indicates support for carryless multiplication.
 #define ARMV8_PMULL (1 << 5)
 
+// ARMV8_SHA512 indicates support for hardware SHA-512 instructions.
+#define ARMV8_SHA512 (1 << 6)
+
 #if defined(__ASSEMBLER__)
 
+// We require the ARM assembler provide |__ARM_ARCH| from Arm C Language
+// Extensions (ACLE). This is supported in GCC 4.8+ and Clang 3.2+. MSVC does
```
Comment (dkostic) [inline include/openssl/arm_arch.h:84]: is this going to be a problem for us? do we use MSVC in the CI or clang? (and what about the infamous gcc4.1?

Comment (bryce-shang) [reply include/openssl/arm_arch.h:84]: MSVC is used in aws-lc-ci-windows-x86.

File: crypto/internal.h (hunk 3 above)
Comment (nebeid) [reply crypto/internal.h]: Done. Removed the functions. Thanks, Dusan.

**Review by dkostic — APPROVED** [review]

**Review by bryce-shang — APPROVED** [review]

**Pull request merged**